ocean_lib/services/v4/service.py

```diff
@@ -0,0 +1,159 @@
+                self.name = service_to_default_name[service_type]
+
+    @classmethod
+    def from_dict(cls, service_dict: Dict[str, Any]) -> "V4Service":
+        """Create a service object from a JSON string."""
+        sd = copy.deepcopy(service_dict)
+        service_id = sd.pop("serviceId", None)
+        service_type = sd.pop("type", None)
+        service_endpoint = sd.pop("serviceEndpoint", None)
+        data_token = sd.pop("datatokenAddress", None)
+        service_files = sd.pop("files", None)
+        timeout = sd.pop("timeout", None)
+        name = sd.pop("name", None)
+        description = sd.pop("description", None)
+        if not service_type:
+            logger.error(
+                'Service definition in DDO document is missing the "type" key/value.'
+            )
+            raise IndexError
+
+        return cls(
+            service_id,
+            service_type,
+            service_endpoint,
+            data_token,
+            service_files,
+            timeout,
+            sd,
+            name,
+            description,
+        )
+
+    @classmethod
+    def from_json(cls, service_dict: Dict[str, Any]) -> "V4Service":
+        return cls.from_dict(service_dict)
+
+    def values(self) -> Dict[str, Any]:
+        """
+
+        :return: array of values
+        """
+        return dict()
+
+    def get_trusted_algos_v4(self) -> list:
+        return self.compute_values.get("publisherTrustedAlgorithms", [])
+
+    def get_trusted_algos_publishers_v4(self) -> list:
+        return self.compute_values.get("publisherTrustedAlgorithmPublishers", [])
+
+    def add_trusted_algo_publisher_v4(self, new_publisher_address: str) -> list:
+        trusted_algo_publishers = [
+            tp.lower() for tp in self.get_trusted_algos_publishers_v4()
+        ]
+        publisher_address = new_publisher_address.lower()
+
+        if publisher_address in trusted_algo_publishers:
+            return trusted_algo_publishers
+
+        trusted_algo_publishers.append(publisher_address)
+        initial_len = len(trusted_algo_publishers)
+        # update with the new list
+        self.compute_values[
+            "publisherTrustedAlgorithmPublishers"
+        ] = trusted_algo_publishers
+        assert (
+            len(self.compute_values["publisherTrustedAlgorithmPublishers"])
+            > initial_len
+        ), "New trusted algorithm was not added. Failed when updating the privacy key. "
+        return trusted_algo_publishers
+
+    def as_dictionary(self) -> Dict[str, Any]:
+        """Return the service as a python dictionary."""
+
+        values = {
+            "serviceId": self.id,
+            "type": self.type,
+            "files": self.files,
+            "datatokenAddress": self.data_token,
+            "serviceEndpoint": self.service_endpoint,
+            "timeout": self.timeout,
+        }
+
+        if self.type == "compute":
+            if "compute" in self.compute_values:
```

Comment:
???

Reply:
I verify if the type of the service is "compute" and after that if the "compute" key is already there. If it's already there I update the compute values, otherwise I'll add the `compute` key. I had some issues if I didn't use `update`, because `compute` key appeared twice like `compute: { compute: {`.